cura/Machines/ContainerTree.py

```diff
@@ -76,54 +74,82 @@ def getCurrentQualityChangesGroups(self) -> List["QualityChangesGroup"]:
+        #   \param definition_id The definition to look for.
+        #   \return A machine node for that definition.
+        def __getitem__(self, definition_id: str) -> MachineNode:
+            if definition_id not in self._machines:
+                start_time = time.time()
+                self._machines[definition_id] = MachineNode(definition_id)
+                self._machines[definition_id].materialsChanged.connect(ContainerTree.getInstance().materialsChanged)
+                Logger.log("d", "Adding container tree for {definition_id} took {duration} seconds.".format(definition_id = definition_id, duration = time.time() - start_time))
+            return self._machines[definition_id]
+
+        ##  Gets a machine node for the specified definition ID, with default.
+        #
+        #   The default is returned if there is no definition with the specified
+        #   ID. If the machine node wasn't loaded yet, this will load it lazily.
+        #   \param definition_id The definition to look for.
+        #   \param default The machine node to return if there is no machine
+        #   with that definition (can be ``None`` optionally or if not
+        #   provided).
+        #   \return A machine node for that definition, or the default if there
+        #   is no definition with the provided definition_id.
+        def get(self, definition_id: str, default: Optional[MachineNode] = None) -> Optional[MachineNode]:
+            if definition_id not in self:
+                return default
+            return self[definition_id]
+
+        ##  Returns whether we've already cached this definition's node.
+        #   \param definition_id The definition that we may have cached.
+        #   \return ``True`` if it's cached.
+        def is_loaded(self, definition_id: str) -> bool:
+            return definition_id in self._machines
+
+    ##  Pre-loads all currently added printers as a background task so that
+    #   switching printers in the interface is faster.
+    class MachineNodeLoadJob(Job):
+        ##  Creates a new background task.
+        #   \param tree_root The container tree instance. This cannot be
+        #   obtained through the singleton static function since the instance
+        #   may not yet be constructed completely.
+        #   \param container_stacks All of the stacks to pre-load the container
+        #   trees for. This needs to be provided from here because the stacks
+        #   need to be constructed on the main thread because they are QObject.
+        def __init__(self, tree_root: "ContainerTree", container_stacks: List["ContainerStack"]):
+            self.tree_root = tree_root
+            self.container_stacks = container_stacks
+            super().__init__()
+
+        ##  Starts the background task.
+        #
+        #   The ``JobQueue`` will schedule this on a different thread.
+        def run(self) -> None:
+            for stack in self.container_stacks:
+                time.sleep(0.5)
```

Comment:
We discussed this (eg; time sleep 0 didn't cause enough of a delay for a thread switch), so it would be nice to document that (Because we usually use time.sleep(0))

Reply:
Done: 5b70d409eda079700485c10b4e32bb4224ed45de